pkg/localstorev2/localstore_test.go

```diff
@@ -0,0 +1,508 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package localstore_test
+
+import (
+	"context"
+	"errors"
+	"io/ioutil"
+	"os"
+	"testing"
+
+	localstore "github.com/ethersphere/bee/pkg/localstorev2"
+	pinstore "github.com/ethersphere/bee/pkg/localstorev2/internal/pinning"
+	"github.com/ethersphere/bee/pkg/localstorev2/internal/upload"
+	chunktesting "github.com/ethersphere/bee/pkg/storage/testing"
+	storage "github.com/ethersphere/bee/pkg/storagev2"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+func verifySessionInfo(
+	t *testing.T,
+	repo *storage.Repository,
+	sessionID uint64,
+	chunks []swarm.Chunk,
+	has bool,
+) {
+	t.Helper()
+
+	for _, ch := range chunks {
+		hasFound, err := repo.ChunkStore().Has(context.TODO(), ch.Address())
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if hasFound != has {
+			t.Fatalf("unexpected chunk state, exp has chunk %t got %t", has, hasFound)
+		}
+	}
+
+	if has {
+		tagInfo, err := upload.GetTagInfo(repo.IndexStore(), sessionID)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if tagInfo.Split != uint64(len(chunks)) {
+			t.Fatalf("unexpected split chunk count in tag, exp %d found %d", len(chunks), tagInfo.Split)
+		}
+		if tagInfo.Seen != 0 {
+			t.Fatalf("unexpected seen chunk count in tag, exp %d found %d", len(chunks), tagInfo.Seen)
+		}
+	}
+}
+
+func verifyPinCollection(
+	t *testing.T,
+	repo *storage.Repository,
+	root swarm.Chunk,
+	chunks []swarm.Chunk,
+	has bool,
+) {
+	t.Helper()
+
+	hasFound, err := pinstore.HasPin(repo.IndexStore(), root.Address())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if hasFound != has {
+		t.Fatalf("unexpected pin collection state, exp exists %t got %t", has, hasFound)
+	}
+
+	allChunks := append([]swarm.Chunk{root}, chunks...)
+
+	for _, ch := range allChunks {
+		hasFound, err := repo.ChunkStore().Has(context.TODO(), ch.Address())
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if hasFound != has {
+			t.Fatalf("unexpected chunk state, exp has chunk %t got %t", has, hasFound)
+		}
+	}
+}
+
+func testUploadStore(t *testing.T, newLocalstore func() (*localstore.DB, error)) {
+	t.Helper()
+
+	t.Run("new session", func(t *testing.T) {
+		t.Parallel()
+
+		lstore, err := newLocalstore()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		for i := 1; i < 5; i++ {
+			id, err := lstore.NewSession()
+			if err != nil {
+				t.Fatal(err)
+			}
+			if id != uint64(i) {
+				t.Fatalf("incorrect id generated, exp %d found %d", i, id)
+			}
+		}
+	})
+
+	t.Run("error on no tag", func(t *testing.T) {
+		t.Parallel()
+
+		lstore, err := newLocalstore()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		_, err = lstore.Upload(context.TODO(), false, 0)
+		if err == nil {
+			t.Fatal("expected error")
+		}
+	})
+
+	for _, tc := range []struct {
+		name   string
+		root   swarm.Chunk
+		chunks []swarm.Chunk
```

Comment:
I would recommend having only the `chunks` field and making the first generated chunk always a root chunk, then you can remove the test `name` and `root` fields. The same goes for the `testPinStore`.